webrtc/videoroom/assets/src/membraneWebRTC.ts

```diff
@@ -101,12 +123,84 @@ export class MembraneWebRTC {
     this.connection = undefined;
   };
 
+  public startScreensharing = async () => {
+    try {
+      this.localScreensharingStream = await navigator.mediaDevices.getDisplayMedia(
+        SCREENSHARING_CONSTRAINTS
+      );
+
+      this.channel
+        .push("start_screensharing", {})
+        .receive("ok", (response) => {
+          this.replaceFakeStreamWithScreenSharing();
+        })
+        .receive("error", (data) => {
+          console.error("error while trying to start screensharing", data);
+          this.localScreensharingStream?.getTracks().forEach((t) => t.stop());
+          this.localScreensharingStream = undefined;
+        });
+    } catch (error) {
+      console.error(error);
+    }
+  };
+
+  public stopScreensharing = async () => {
+    this.localScreensharingStream?.getTracks().forEach((t) => {
+      t.stop();
+      // on why we dispatch 'ended' manually see: https://stackoverflow.com/questions/55953038/why-is-the-ended-event-not-firing-for-this-mediastreamtrack
+      t.dispatchEvent(new Event("ended"));
+    });
+  };
+
+  private replaceFakeStreamWithScreenSharing = () => {
+    const [fakeTrack] = this.fakeStream.getTracks();
+    const screenSender = this.connection!.getSenders().find((sender) => {
+      return sender?.track?.id === fakeTrack.id;
+    });
+
+    if (!screenSender) {
+      console.error("Could not find sender track for", fakeTrack);
+      this.channel.push("stop_screensharing", {});
+      return;
+    }
+
+    screenSender.replaceTrack(this.localScreensharingStream?.getTracks()[0]!);
+
+    this.localScreensharingStream!.getTracks().forEach((t) => {
+      t.onended = (_) => {
+        screenSender.replaceTrack(fakeTrack);
+        this.callbacks.onScreensharingEnd?.();
+        this.channel.push("stop_screensharing", {});
+        this.localScreensharingStream = undefined;
+      };
+    });
+
+    this.callbacks.onScreensharingStart?.(this.localScreensharingStream!, true);
+  };
+
   private onOffer = async (offer: OfferData) => {
     if (!this.connection) {
       this.connection = new RTCPeerConnection(this.rtc_config);
       this.connection.onicecandidate = this.onLocalCandidate();
       this.connection.ontrack = this.onTrack();
-      this.localTracks.forEach(track => this.connection!.addTrack(track));
+      this.localTracks.forEach((track) =>
+        this.connection!.addTrack(track, this.localStream!)
+      );
+
+      this.connection.oniceconnectionstatechange = (event) => {
+        console.log(
+          "Ice connection state:",
+          (event.target as RTCPeerConnection).iceConnectionState
+        );
+        if (
+          (event.target as RTCPeerConnection).iceConnectionState === "connected"
+        ) {
+          this.channel.push("connected", {});
+        }
+      };
+      this.fakeStream
+        .getTracks()
+        .forEach((track) => this.connection!.addTrack(track, this.fakeStream));
```

Comment:
I wouldn't add any fake streams by default. Instead, we could have `addFakeTrack` (as an alias for `addTrack(fakeTrack, ...)`) and `replaceTrack`, to stay close to the WebRTC API

Reply:
I don't know if I understand you correctly, but it is necessary for a stream to be there when negotiation is being performed. We have to provide fake stream as we don't support dynamic adding/removing of tracks in our pipeline.